---
title: "PUBLIC HEALTH WHOLE GENOME SEQUENCING ANALYSIS REPORT"
output:
  html_document:
    theme: paper
    df_print: paged
---

```{css, echo=FALSE}
@import url('https://fonts.googleapis.com/css2?family=Noto+Sans+JP:wght@100;400;500;700&family=Roboto:wght@100;300;400;500;700&display=swap');
body {
  font-family: 'Noto Sans JP', Arial, Helvetica, sans-serif;
  font-size: 2rem;
  background-color: #FFF;
}

.title {
  font-weight: 300;
  text-align: center;
  margin-bottom: 50px;
  margin-top: 160px;
  font-size: 6rem !important;
}

.subtitle {
  font-weight: 300;
  text-align: center;
  font-size: 3rem !important;
}

h1, h2 {
  font-family: "Roboto", Arial, Helvetica, sans-serif;
  margin: 0px;
  font-weight: 700;
}

h3, h4, h5, h6 {
  font-family: "Roboto", Arial, Helvetica, sans-serif;
  margin: 10px;
  font-weight: 500;
}

.main-container{
  width: 85%;
  max-width: none;
  background-color: #FFF;
}
.section {
  margin-bottom: 50px;
}
h1 { font-size: 3.4rem;}
h2 { font-size: 3rem; }
h3 { font-size: 2.6rem; }
h4 { font-size: 2.2rem; }
h5 { font-size: 2rem; }

h2:before,
h2::after {
    display: inline-block;
    content: "";
    border-top: .3rem solid;
    width: 4rem;
    margin: 0 1rem;
    transform: translateY(-1rem);
}

.titlePanel {
  background-color: #F0F0F0;
  position:absolute;
  top:0px;
  height: 135px;
  width: 100%;
  left:0px;
}
```

<div class ="titlePanel">
```{r, echo=FALSE, fig.align='left',out.height='100px',out.extra="style='margin:20px'"}
knitr::include_graphics(config$logo)
```
```{r, echo=FALSE, fig.align='right',out.height='100px',out.extra="style='margin:20px'"}
knitr::include_graphics(config$logo2)
```
</div>

```{r setup, include=FALSE}

#if (!requireNamespace("BiocManager", quietly = TRUE))
#  install.packages("BiocManager")
#BiocManager::install("ggtree")

library(ggplot2)
library(plotly)
library(pander)
library(kableExtra)
library(ape)
library(ggtree)
library(heatmaply)
library(phytools)

knitr::opts_chunk$set(echo = FALSE,warning=FALSE,message=FALSE)

```

<div class="subtitle">
`r subHeaderText`
</div>

<center>
```{r intro}

# create table
colnames(headerDF) <- c("REPORT DATE","PROJECT NAME","PREPARED BY")

# plot table
pander(headerDF)
```
</center>

## SAMPLE INFO  {.tabset}
`r summaryTEXT`

### QC {.tabset}
<center>
```{r qc}
# https://github.com/CDCgov/phoenix/wiki/Pipeline-Overview#pipeline-qc-checks
kbl(sampleDF[,c(1:3,8:ncol(sampleDF))]) %>%
  kable_styling(bootstrap_options = c("hover", "condensed", "responsive"),
  fixed_thead = T,
  font_size = 14)
```
</center>

### SUMMARY {.tabset}

<center>
```{r samples}
# plot table
kbl(sampleDF[,1:6]) %>%
  kable_styling(bootstrap_options = c("hover", "condensed", "responsive"),
  fixed_thead = T,
  font_size = 14)
```
</center>

## RELATEDNESS {.tabset}

### HEATMAP {.tabset}
`r config$heatmap.text`
<center>
```{r heatmap}
if(exists('snpData') && config$show.snp==TRUE){
  heatmaply(snpData,
            cellnote=snpData,
            cellnote_textposition = "middle center",
            colors = viridis(n = 256, alpha = 1, begin = 1, end = 0, option = "viridis"),
            show_dendrogram = c(FALSE, FALSE),
            dist_method=config$heat.dist.method, label_names = c("Row", "Column", "SNPs"))
}

if(exists('snpData') && config$show.snp==FALSE){
  rownames(snpData) <- snpData$X
  snpData <- snpData [,-1]
  heatmaply(snpData,
            colors = viridis(n = 256, alpha = 1, begin = 1, end = 0, option = "viridis"),
            show_dendrogram = c(FALSE, FALSE),
            dist_method=config$heat.dist.method)
}

```
</center>
***
### PHYLOGENETIC TREE {.tabset}
`r config$tree.text`
`r if(exists('treepath') & exists('cgstats')){paste("Core Genes Identified: ",cgstats[cgstats$V1 == 'Core genes',3],'\n')}`
`r if(exists('treepath') & exists('cgstats')){paste("Total Genes Identified: ",cgstats[cgstats$V1 == 'Total genes',3])}`

<center>
```{r tree, message=FALSE}
if(exists('treepath')){
  set.seed(42)
  tree <- ape::read.tree(treepath)
  if(exists('cgstats')){
    paste("Core Genes Identified: ",cgstats[cgstats$V1 == 'Core genes',3])
    paste("Total Genes Identified: ",cgstats[cgstats$V1 == 'Total genes',3])
  }

  if(config$root.method == 'midpoint'){
    tree <- midpoint.root(tree)
  }
  else if(config$root.method == 'unrooted'){
    tree <- tree
  }
  else {
    node <- match(config$root.method, tree$tip.label)
    if(is.na(node)){
      message('Root sample ID not found')
      quit(save="no", status=1)
    } else {
      tree <- reroot(tree,node)
    }
  }

  # metadata
  id <- tree$tip.label

  # group samples if we have a grouping catagory
  if(any(!is.na(sampleDF$TreeGroup))){
    #reorder sampleDF
    orderedDF <- sampleDF[match(id,sampleDF$'Lab ID'),]
    grp <- orderedDF$TreeGroup
  }
  else {
    grp <- NULL
  }

  dat <- tibble::tibble(id = id)
  # plot tree
  ggtree_plot <- ggtree(tree)
  metat <- ggtree_plot$data %>%
    dplyr::inner_join(dat, c('label' = 'id'))

  ggtree_plot <- ggtree_plot +
    geom_point(data = metat,
               aes(x = x,
                   y = y,
                   colour = grp,
                   label = id, text = paste0(id)),show.legend = FALSE)


  if(config$show.bootstrap){
     #filter bootstrap values
     d <- ggtree_plot$data
     bs <- d[!d$isTip,]
     bs$label <- as.numeric(bs$label)
     bs <- bs[bs$label > config$bootstrap.threshold,]
     ggtree_plot <- ggtree_plot + geom_text(data=bs,aes(label=label,),position = position_dodge2(width = 0.001))
  }

  ggplotly(ggtree_plot, tooltip = "text")
}
```
</center>

***
### AR GENE SUMMARY {.tabset}
`r config$ar_gene_summary.text`
<center>
```{r ar_table}
if(exists('ar_summary')){
  
  # determine gene consensus at 50%
  number_of_samples=round(length(unique(ar_summary$Sample))*.5+.5,0)
  output_df=data.frame()
  for (geneID in unique(ar_summary$Gene)){
    tmp_df=subset(ar_summary,Gene==geneID)
    if (nrow(tmp_df)==number_of_samples){
      if (nrow(output_df)==0){
        output_df=tmp_df
      } else{
        output_df=rbind(tmp_df,output_df)
      }
    }
  }
  # strip rownames, order by gene name
  row.names(output_df) <- NULL
  output_df <- output_df[order(output_df$Gene),]
  
  # if there are consensus genes then print table
  if (nrow(output_df)==0){
    print("There are no genes with consensus between at least 50% of samples")
  } else{
    kbl(output_df[,c(1,3,4)])  %>%
      kable_styling(bootstrap_options = c("hover", "condensed", "responsive"),
      fixed_thead = T,
      font_size = 14) %>%
      pack_rows(index=table(output_df$Gene))
  }
}
```
</center>

`r if(exists('matrixpng')){"## PANGENOME ANALYSES {.tabset}"}`
`r if(exists('matrixpng')){config$matrix.text}`

`r if(exists('piepng')){"### PANGENOME (PRESENT/ABSENCE) {.tabset}"}`
`r if(exists('piepng')){config$pangenome.pa}`
<center>
```{r pangenome1}
if(exists('matrixpng')){
  knitr::include_graphics(matrixpng)
}
```
</center>

`r if(exists('piepng')){"### PANGENOME (PIE) {.tabset}"}`
`r if(exists('piepng')){config$pangenome.pie}`

<center>
```{r pangenome2}
if(exists('piepng')){
  knitr::include_graphics(piepng)
}
```
</center>

`r if(exists('freqpng')){"### PANGENOME (FREQUENCY) {.tabset}"}`
`r if(exists('freqpng')){config$pangenome.frequency}`

<center>
```{r pangenome3}
if(exists('piepng')){
  knitr::include_graphics(freqpng)
}
```
</center>

## METHODS
`r methodsTEXT`

## DISCLAIMER
`r disclaimerTEXT`
