---
title: "Public Health Whole Genome Sequencing Analysis Report"
output: 
  flexdashboard::flex_dashboard:
    orientation: columns
    vertical_layout: fill
    navbar:
      - { title: "ODHL", href:
      "https://odh.ohio.gov/know-our-programs/microbiology/microbiology",
      align_right}
    logo: "REP_LOGO"
params:
  configFILE: "REP_CONFIG"
  projectname: "REP_PROJID"
  outdir: "REP_OUT"
  date: "REP_DATE"
  sampletable: "REP_ST"
  snpmatrix: "REP_SNP"
  tree: "REP_TREE"
  cgstats: "REP_CORE"
  arpredictions: "REP_AR"
  metadata: "REP_META"
  testing: "N"
  obid: "REP_OB"
editor_options: 
  chunk_output_type: console
---

```{r libs, echo=FALSE,include=FALSE,warning=FALSE,message=FALSE}
library(rmarkdown)
library(argparser)
library(yaml)
library(ape)
library(ggplot2)
library(plotly)
library(pander)
library(kableExtra)
library(ape)
library(ggtree)
library(heatmaply)
library(phytools)
library(DT)
library(RColorBrewer)
library(ggpubr)
library(flexdashboard)
library(crosstalk)
require(phytools)
require(rgl)
require(htmlwidgets)
knitr::opts_chunk$set(echo = FALSE,warning=FALSE,message=FALSE)
options(rgl.useNULL = TRUE) # Suppress the separate window.
'%ni%' <- function(x,y)!('%in%'(x,y))
```

```{r args, echo=FALSE,warning=FALSE}
username="S. Chill"
reporttype="Outbreak"

#position args
if (params$testing=="N"){
  projectname=params$projectname
  outdir=params$outdir
  configFILE=params$configFILE
  date=params$date
  sampletable=params$sampletable
  snpmatrix=params$snpmatrix
  tree=params$tree
  cgstats=params$cgstats
  arpredictions=params$arpredictions
  metadata=params$metadata
  obid=params$obid
} else {
  projectname="OH-M6588-231201"
  outdir="OH-M6588-231201/"
  date="12/18/23"
  sampletable="final_report.csv"
  snpmatrix="snp_distance_matrix.tsv"
  tree="core_genome.tree"
  cgstats="core_genome_statistics.txt"
  arpredictions="ar_predictions.tsv"
  metadata="manifest.csv"
}
```

```{r prep, echo=FALSE, warning=FALSE}
# read yaml file
config <- read_yaml(configFILE)

## set header text
subHeaderText = paste0(config$sub.title, ": ", reporttype)

## get header table
headerDF <- data.frame(date=date,ob=obid, project=projectname,name=username)

## get summary text
summaryTEXT <- config$summary.paragraph

## get disclaimer text
disclaimerTEXT <- config$disclaimer.text

## get methods text
methodsTEXT <- config$methods.text

## get reference
referenceTEXT <- config$reference.text
```

```{r read, echo=FALSE}
## get sample table
sampleDF <- read.csv2(sampletable,sep=",")
sampleDF=subset(sampleDF,species!="species") #handles merged proj dups
rownames(sampleDF)=sampleDF$specimen_id

# get metadata
metadataDF = read.csv(metadata)

# sampleDF and metadata  merge
sampleDF=merge(sampleDF,metadataDF[,c("specimen_id","pt_city","isolation_source",
                                      "pt_county","healthcare_origin",
                                      "submitter_name","collect_date")], 
               by="specimen_id",all=TRUE)

# replace NAs
sampleDF$srr_id <- ifelse(sampleDF$srr_id == "NO_ID", sampleDF$specimen_id,
                          sampleDF$srr_id)
sampleDF$pt_city[is.na(sampleDF$pt_city)] <- "Unknown"
sampleDF$healthcare_origin[is.na(sampleDF$healthcare_origin)] <- "Unknown"

# get heatmap
snpData <- read.csv2(snpmatrix,sep='\t',check.names = F,row.names = 1)

# get tree
treepath <- tree

# get cgstats
cgstats <- read.csv2(cgstats,sep='\t',header = FALSE)

# get ar-summary
ar_summary <- read.csv2(arpredictions,sep='\t',check.names = F)
ar_summary$Sample=gsub("_[0-9][0-9][0-9][0-9]_length_[0-9]*","",ar_summary$Sample)
ar_summary$Sample=gsub("_[0-9][0-9][0-9]_length_[0-9]*","",ar_summary$Sample)
ar_summary$Sample=gsub("_[0-9][0-9]_length_[0-9]*","",ar_summary$Sample)
ar_summary$Sample=gsub("_[0-9]_length_[0-9]*","",ar_summary$Sample)
ar_summary=ar_summary[!duplicated(ar_summary),]

# create metadata
generate_ids=function(list_in){
  tmp_list=which( colnames(sampleDF) %ni% list_in ) -1
  return(tmp_list)
}
id_nums=generate_ids(c("auto_qc_outcome","specimen_id","wgs_id","srr_id"))
tax_nums=generate_ids(c("specimen_id","species","sequence_classification",
                        "mlst_1","gamma_beta_lactam_resistance_genes"))
qc_nums=generate_ids(c("specimen_id","estimated_coverage","genome_length",
                        "auto_qc_failure_reason"))
sample_meta=SharedData$new(sampleDF)
```

```{r summarytext, echo=FALSE}
# generate summary test for outupt
pass_n=nrow(subset(sampleDF, auto_qc_outcome=="PASS"))
failed_n=nrow(subset(sampleDF, auto_qc_outcome!="PASS"))
failed_samples=subset(sampleDF, auto_qc_outcome!="PASS")$specimen_id
summaryTEXT=paste0("There were ", nrow(sampleDF)," samples analyzed and ",
                   pass_n, " samples passed quality control thresholds, while ",
                   length(failed_samples)," failed.")
if(failed_n>0){
  summaryTEXT=paste0(summaryTEXT,
                     " Failing samples included: ",
                     paste(failed_samples,collapse=" | "),".")
}
```

Information {data-orientation=columns data-icon="fa-info-circle"}
=================================================================
Column {data-width=600}
-----------------------
### `r subHeaderText` 
```{r intro}
# create table
colnames(headerDF) <- c("REPORT DATE","OUTBREAK ID","PROJECT NAME","PREPARED BY")

# plot table
pander(headerDF)
```

### METHODS
`r methodsTEXT`
Column {data-width=600}
-----------------------
### Summary
`r summaryTEXT`

`r referenceTEXT`

### DISCLAIMER
`r disclaimerTEXT`

Datatables {data-orientation=columns data-icon="ion-search"}
=================================================================

Input {.sidebar}
-----------------------------------------------------------------------

### Filters 
```{r filter}
filter_select(
  id="auto_qc_outcome",
  label = "QC Status",
  sharedData = sample_meta,
  group = ~auto_qc_outcome
)
filter_select(
  id="PT County",
  label = "pt_county",
  sharedData = sample_meta,
  group = ~pt_county
)
filter_select(
  id="healthcare_origin",
  label = "Healthcare Origin",
  sharedData = sample_meta,
  group = ~healthcare_origin
)
```

Raw {data-width=700 .tabset}
-----------------------------------------------------------------------
### Sample

```{r ids, echo=FALSE}
sample_meta %>%
  DT::datatable(
    extensions = c("Buttons","Scroller"),
    rownames=FALSE,
    style="bootstrap",
    width="100%",
    class="compact",
    options=list(
      dom="Blrtrip",
      deferREander=TRUE,
      columnDefs=list(
        list(
          visible=FALSE,
          targets=id_nums
        )
      ),
      buttons = list(I("colvis"),"csv","excel")
      ),
    colnames = c(
      "QC" = "auto_qc_outcome",
      "eLIMS_ID" = "specimen_id",
      "WGS" = "wgs_id",
      "SRR" = "srr_id",
      "Lab_Results" = "lab_results"
    )
  )
```

### Taxonomy
```{r tax, echo=FALSE}
sample_meta %>%
  DT::datatable(
    extensions = c("Buttons","Scroller"),
    rownames=FALSE,
    style="bootstrap",
    width="100%",
    class="compact",
    options=list(
      dom="Blrtrip",
      deferREander=TRUE,
      columnDefs=list(
        list(
          visible=FALSE,
          targets=tax_nums
        )
      ),
      buttons = list(I("colvis"),"csv","excel")
      ),
    colnames = c(
      "eLIMS_ID" = "specimen_id",
      "Species" = "species",
      "Sequence_Classification" = "sequence_classification",
      "MLST1" = "mlst_1",
      "Beta-lactamase resistant genes" = "gamma_beta_lactam_resistance_genes"
    )
  )
```

```{r ar_table, echo=FALSE}
### ARGene Table
if(exists('ar_summary')){
  
  # determine gene consensus at 50%
  number_of_samples=round(length(unique(ar_summary$Sample))*.5+.5,0)
  output_df=data.frame()
  for (geneID in unique(ar_summary$Gene)){
    tmp_df=subset(ar_summary,Gene==geneID)
    if (nrow(tmp_df)>number_of_samples-1){
      if (nrow(output_df)==0){
        output_df=tmp_df
      } else{
        output_df=rbind(tmp_df,output_df)
      }
    }
  }
  # strip rownames, order by gene name
  row.names(output_df) <- NULL
  output_df <- output_df[order(output_df$Gene),]

}
```

Analysis {data-orientation=columns data-icon="ion-android-options"}
=================================================================
Column {data-width=800 .tabset}
-----------------------------------------------------------------------
### Taxonomy
```{r counts}
subset(sampleDF,auto_qc_outcome=="PASS")%>%
  ggplot(aes(x = species,fill=species)) +
  geom_bar() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```

### Heatmap
```{r heatmap}
collabels=as.data.frame(colnames(snpData))
colnames(collabels)="specimen_id"
collabels=merge(collabels,sampleDF[,c("specimen_id","healthcare_origin")])

rowlabels=as.data.frame(rownames(snpData))
colnames(rowlabels)="specimen_id"
rowlabels=merge(rowlabels,sampleDF[,c("specimen_id","mlst_1")])

heatmaply(snpData,
          cellnote=snpData,
          cellnote_size=8,
          cellnote_textposition = "middle center",
          col_side_colors = data.frame("healthcare_origin" = collabels[,c("healthcare_origin")],
                                       check.names=FALSE),
          row_side_colors = data.frame("MLST" = rowlabels[,c("mlst_1")],
                                       check.names=FALSE),
          colors = viridis(n = 256, alpha = 1, 
                           begin = 1, end = 0, option = "viridis"),
          show_dendrogram = c(TRUE, FALSE),
          dist_method=config$heat.dist.method, 
          label_names = c("Row", "Column", "SNPs"))

```

### Phylogenetic Tree

```{r treeparams}
if(exists('treepath')){
  set.seed(42)
  tree <- ape::read.tree(treepath)
  if(exists('cgstats')){
    paste("Core Genes Identified: ",cgstats[cgstats$V1 == 'Core genes',3])
    paste("Total Genes Identified: ",cgstats[cgstats$V1 == 'Total genes',3])
  }

  # set the root method
  if(config$root.method == 'midpoint'){
    tree <- midpoint.root(tree)
  } else if(config$root.method == 'unrooted'){
    tree <- tree
  } else {
    node <- match(config$root.method, tree$tip.label)
    if(is.na(node)){
      message('Root sample ID not found')
      quit(save="no", status=1)
    } else {
      tree <- reroot(tree,node)
    }
  }
}
```

#### Ancestrial Tree

A phylogeny tree is used to illustrate relationships where a simulation of continuous trait evolution is illustrated (Brownian motion BM model). The "trait value" refers to the values of the simulated continuous trait for each tip (or node). The length of each branch is representative of lineage relatedness. The first image includes all samples; all subsequent images shows subsets of the complete dataset, highlighting other potential relationships.

1. Complete dataset
2. Removing all samples with "Unknown" healthcare origin"
3. Removing subset of samples clustering most differently

```{r treepruned}
x<-fastBM(tree)
obj<-contMap(tree,x)
plot(obj,type="fan",lwd=2,fsize=c(.8,.5))

# plot3
tree_pruned <- drop.tip(tree, 
                        subset(sampleDF,healthcare_origin=="Unknown")$specimen_id)
x_pruned<-fastBM(tree_pruned)
obj_pruned<-contMap(tree_pruned,x_pruned)
plot(obj_pruned,type="fan",lwd=2,fsize=c(.7,.5)) ## for example

# plot3
```


#### Cladogram

A cladogram is used to illustrate relationships, where nodes are labeled with degree of similarity, but where the vertical lines are not representative of lineage relatedness.
```{r tree, message=FALSE}
#https://yulab-smu.top/treedata-book/chapter9.html
# create basic tree with equal lengths
p=ggtree(tree,branch.length="none")+  layout_dendrogram()
strip=ggtree(tree,branch.length="none")

# pull data to create node labels
d <- p$data
bs <- d[!d$isTip,]
bs$label <- as.numeric(bs$label)
ggtree_plot <- p + 
  geom_text(data=bs,aes(label=label,vjust=1.2),size=2)

# create df for labels
rownames(sampleDF)=sampleDF$specimen_id
sampleDF$ref <- as.integer(factor(sampleDF$pt_city, levels = unique(sampleDF$pt_city)))
labels=p$data$label[p$data$label %ni% bs$label]
d <- data.frame(label = labels, 
                  ref  = sampleDF[labels, c("ref")],
                  mlst = sampleDF[labels, c("mlst_1")])

# plot graph
ggtree_plot %<+% d + 
  layout_dendrogram()+  
  geom_tippoint(aes(fill=factor(mlst), x=x+.5), 
                size=6, shape=21, color='black')+
  geom_tiplab(aes(label=ref), size=2.5, offset=-.4,color='black')+
  geom_tiplab(hjust=1, offset=-1.5, show.legend=FALSE)+
  theme_dendrogram(plot.margin=margin(0,0,80,0))
```

#### Morphospace
Expanding on the evolutionary relatedness, a multi-dimensional space is created to illustrated how closely related species tend to cluster together in morphospace. This reflects the phylogenetic signal in the morphological data, showing how traits have evolved in a manner that is correlated with the species' evolutionary history.

```{r treemorph}
tree_morph=ape::read.tree(treepath)
ids=tree_morph$tip.label
DT::datatable(data.frame(ID=c(1:length(ids)),
                         wgs_ID=ids,
                         HealthcareOrigin=sampleDF[ids,"healthcare_origin"]))
tree_morph$tip.label=c(1:length(ids))
X<-fastBM(tree_morph,nsim=3) ## simulate 3 characters
colors<-sapply(tree$edge[,2], function(x,n) if(x>n) "blue"
  else "red",n=length(tree$tip.label))
phylomorphospace3d(tree_morph,X,control=list(col.edge=colors))
rglwidget()
```

### ARGene Summary
```{r ar_summary}
if(exists('ar_summary')){
  # if there are consensus genes then print table
  if (nrow(output_df)==0){
    print("There are no genes with consensus between at least 50% of samples")
  } else{
    merge(sampleDF,output_df, by.x="specimen_id",
      by.y="Sample")[,c("specimen_id","Gene","species","run_id")] %>% 
      ggplot(aes(x = Gene,fill=species)) +
      geom_bar() + 
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
  }
}
```

Column {data-width=300}
-----------------------------------------------------------------------
### Notes

#### Heatmap
`r config$heatmap.outbreak.text`

#### Phylogenetic Tree
`r config$tree.text`
`r if(exists('treepath') & exists('cgstats')){paste("Core Genes Identified: ",cgstats[cgstats$V1 == 'Core genes',3],'\n')}`
`r if(exists('treepath') & exists('cgstats')){paste("Total Genes Identified: ",cgstats[cgstats$V1 == 'Total genes',3])}`

#### ARGene
`r config$ar_gene_summary.text`